File: Array.cpp
#include "Array.h"
#include <iostream>


void initArray(Array& arr) {
    arr.size = 0;
    arr.memorySize = 0;
    arr.memory = new std::string[0];
}

void addArray(Array& arr, const std::string& key) {
    if (arr.size + 1 > arr.memorySize) {
        int newMemorySize = arr.size * 2 + (arr.size == 0);
        std::string* newMemory = new std::string[newMemorySize];
        for (int i = 0; i < arr.size; i++)
            newMemory[i] = arr.memory[i];
        delete[] arr.memory;
        arr.memory = newMemory;
        arr.memorySize = newMemorySize;
    }
    arr.memory[arr.size] = key;
    arr.size++;
}

void addArray(Array& arr, const std::string& key, int index) {
    if (index < 0 || index > arr.size) return;

    if (arr.size + 1 > arr.memorySize) {
        int newMemorySize = arr.size * 2 + (arr.size == 0);
        std::string* newMemory = new std::string[newMemorySize];
        for (int i = 0; i < index; i++)
            newMemory[i] = arr.memory[i];
        newMemory[index] = key;
        for (int i = index + 1; i < arr.size + 1; i++)
            newMemory[i] = arr.memory[i - 1];
        delete[] arr.memory;
        arr.memory = newMemory;
        arr.memorySize = newMemorySize;
    }
    else {
        for (int i = arr.size; i > index; i--)
            arr.memory[i] = arr.memory[i - 1];
        arr.memory[index] = key;
    }
    arr.size++;
}

std::string getArray(Array& arr, int index) {
    if (index < 0 || index >= arr.size) return "";
    return arr.memory[index];
}

void removeArray(Array& arr, int index) {
    if (index < 0 || index >= arr.size) return;
    for (int i = index; i < arr.size - 1; i++)
        arr.memory[i] = arr.memory[i + 1];
    arr.size--;
}

void changeArray(Array& arr, int index, const std::string& key) {
    if (index < 0 || index >= arr.size) return;
    arr.memory[index] = key;
}

int lengthArray(Array& arr) {
    return arr.size;
}

void printArray(Array& arr) {
    for (int i = 0; i < arr.size; i++)
        std::cout << arr.memory[i] << ' ';
        std::cout << std::endl;
}

void deleteArray(Array& arr) {
    delete[] arr.memory;
    arr.memory = nullptr;
    arr.size = 0;
    arr.memorySize = 0;
}

File: golang/array.go
package main

import "fmt"

type Array struct {
	memory []string
	size   int
}

func initArray(arr *Array) {
	arr.memory = []string{}
	arr.size = 0
}

func addArray(arr *Array, value string) {
	arr.memory = append(arr.memory, value)
	arr.size++
}

func addIndexArray(arr *Array, value string, index int) {
	if index < 0 || index > arr.size {
		return
	}
	arr.memory = append(arr.memory[:index], append([]string{value}, arr.memory[index:]...)...)
	arr.size++
}

func getArray(arr *Array, index int) string {
	if index < 0 || index >= arr.size {
		return ""
	}
	return arr.memory[index]
}

func removeArray(arr *Array, index int) {
	if index < 0 || index >= arr.size {
		return
	}
	arr.memory = append(arr.memory[:index], arr.memory[index+1:]...)
	arr.size--
}

func changeArray(arr *Array, index int, value string) {
	if index < 0 || index >= arr.size {
		return
	}
	arr.memory[index] = value
}

func lengthArray(arr *Array) int {
	return arr.size
}

func printArray(arr *Array) {
	for _, v := range arr.memory {
		fmt.Print(v, " ")
	}
	fmt.Println()
}

func deleteArray(arr *Array) {
	arr.memory = nil
	arr.size = 0
}
